code/lib/core-server/src/presets/common-override-preset.ts

```diff
@@ -19,30 +26,51 @@ export const framework: PresetProperty<'framework', StorybookConfig> = async (co
 
 export const stories: PresetProperty<'stories', StorybookConfig> = async (entries, options) => {
   if (options?.build?.test?.disableMDXEntries) {
-    return (
+    const result = (
       await Promise.all(
         normalizeStories(entries, {
           configDir: options.configDir,
           workingDir: options.configDir,
-        }).map(({ directory, files }) => {
+        }).map(async ({ directory, files, titlePrefix }) => {
           const pattern = join(directory, files);
           const absolutePattern = isAbsolute(pattern) ? pattern : join(options.configDir, pattern);
+          const absoluteDirectory = isAbsolute(directory)
+            ? directory
+            : join(options.configDir, directory);
 
-          return glob(slash(absolutePattern), {
-            ...commonGlobOptions(absolutePattern),
-            follow: true,
-          });
+          return {
+            files: (
+              await glob(slash(absolutePattern), {
+                ...commonGlobOptions(absolutePattern),
+                follow: true,
+              })
+            ).map((f) => relative(absoluteDirectory, f)),
+            directory,
+            titlePrefix,
+          };
         })
       )
-    ).flatMap((expanded, i) => {
-      const filteredEntries = expanded.filter((s) => !s.endsWith('.mdx'));
+    ).reduce<StoriesEntry[]>((acc, expanded, i) => {
+      const filteredEntries = expanded.files.filter((s) => !s.endsWith('.mdx'));
       // only return the filtered entries when there is something to filter
       // as webpack is faster with unexpanded globs
-      if (filteredEntries.length < expanded.length) {
-        return filteredEntries;
+      let items = [];
+      if (filteredEntries.length < expanded.files.length) {
+        items = filteredEntries.map((k) => ({
+          ...expanded,
+          files: `**/${k}`,
```

Comment:
This is the part I'm not really happy about... But it was the only solution I could really find without digging into this:

https://github.com/storybookjs/storybook/blob/ff6b28e3915b0505961e61f3268effdc2aa9cc76/code/lib/core-common/src/utils/glob-to-regexp.ts#L3-L27

... and it really do not want to go digging into that code, right now.

I'm thinking this is "close enough" for now.

Reply:
It would be a lot easier to understand what this code was doing if we had a test for it.

I am assuming it turns something like:

```
[{ directory: 'foo', files: '**/*.(mdx|stories.js)' }]
```

Into something like:
```
[
  { directory: 'foo', files: '**/first-story.stories.js' },
  { directory: 'foo', files: '**/second-story.stories.js' },
  //..
]
```

Where that list can be very long? That seems like it would have a pretty detritmental impact on build times if that list was really long, but I guess we would have to test it it out to know for sure.

I wonder if there's a way to hit the most common cases and just drop the `mdx|` bit from the regexp, if we see a common pattern like `'**/*.@(mdx|stories.@(js|jsx|mjs|ts|tsx))';`

Also, what if the `expanded` list is empty?